lib/maths/CTimeSeriesTestForSeasonality.cc

```diff
@@ -0,0 +1,860 @@
+    }
+}
+
+void CTimeSeriesTestForSeasonality::addDecomposition(const TFloatMeanAccumulatorVec& valuesMinusTrend,
+                                                     TFloatMeanAccumulatorVec& valuesToDecompose,
+                                                     std::size_t numberTrendSegments,
+                                                     TSeasonalComponentVec& periods,
+                                                     TModelVec& decompositions) const {
+    valuesToDecompose = valuesMinusTrend;
+    periods = CSignal::seasonalDecomposition(
+        valuesToDecompose, m_OutlierFraction, {this->day(), this->week(), this->year()},
+        [](std::size_t) { return 1.0; }, m_StartOfWeek);
+    this->removeIfInsufficientData(periods);
+
+    if (periods.size() > 0 && this->alreadyModelled(periods) == false &&
+        this->onlyDiurnal(periods) == false) {
+
+        auto decomposition = this->testDecomposition(valuesToDecompose,
+                                                     numberTrendSegments, periods);
+        if (decomposition.seasonal()) {
+            for (std::size_t i = 0; i < m_Values.size(); ++i) {
+                double offset{CBasicStatistics::mean(m_Values[i]) -
+                              CBasicStatistics::mean(valuesMinusTrend[i])};
+                CBasicStatistics::moment<0>(decomposition.s_TrendInitialValues[i]) = offset;
+            }
+            decompositions.push_back(std::move(decomposition));
+        }
+    }
+}
+
+CTimeSeriesTestForSeasonality::SModel
+CTimeSeriesTestForSeasonality::testDecomposition(const TFloatMeanAccumulatorVec& valuesToTest,
+                                                 std::size_t numberTrendSegments,
+                                                 const TSeasonalComponentVec& periods) const {
+    using TComputeScaling =
+        std::function<bool(TFloatMeanAccumulatorVec&, SHypothesisStats&)>;
+
+    LOG_TRACE(<< "testing " << core::CContainerPrinter::print(periods));
+
+    TComputeScaling scalings[]{
+        [&](TFloatMeanAccumulatorVec& values, SHypothesisStats& stats) {
+            stats.s_ScaleSegments.assign({0, values.size()});
+            return true;
+        },
+        [&](TFloatMeanAccumulatorVec& values, SHypothesisStats& hypothesis) {
+            std::size_t period{hypothesis.s_Period.period()};
+            hypothesis.s_ScaleSegments =
+                TSegmentation::piecewiseLinearScaledSeasonal(values, period);
+            return this->meanScale(values, hypothesis,
+                                   [](std::size_t) { return 1.0; });
+        }};
+
+    TFloatMeanAccumulatorVec residuals{valuesToTest};
+    THypothesisStatsVec hypotheses;
+    hypotheses.reserve(periods.size());
+
+    for (std::size_t i = 0; i < periods.size(); ++i) {
+
+        LOG_TRACE(<< "testing " << periods[i].print());
+
+        // Precondition by removing all remaining components.
+        m_ValuesToTestComponent = residuals;
+        m_Periods.clear();
```

Comment:
The mutable state is to avoid repeated allocations of relatively large objects. The rule-of-thumb reading this class is that all mutable state is only used very near where it is initialised, i.e. think of these as memory buffers with which to work.

I did indeed use local variables for everything initially, but this turned up as relatively expensive in the profiling I did. I also tried passing in references to all this state to each function and found that less readable. If you can think of a better alternative which avoids the repeated memory allocations I'd be happy to change, but I couldn't find a cleaner alternative.